README.rst
.. image:: https://travis-ci.org/teadur/whoisd-rwhois.svg?branch=master
    :target: https://travis-ci.org/teadur/whoisd-rwhois   
.. image:: https://api.codeclimate.com/v1/badges/f2f9b7fe3f0367b860a7/maintainability
   :target: https://codeclimate.com/github/teadur/whoisd-rwhois/maintainability
   :alt: Maintainability
.. image:: https://api.codeclimate.com/v1/badges/f2f9b7fe3f0367b860a7/test_coverage
   :target: https://codeclimate.com/github/teadur/whoisd-rwhois/test_coverage
   :alt: Test Coverage

=============
whoisd-rwhois
=============


Unix whois daemon for restful whois backend.


python src/whoisd_rwhois/whoisd.py -vv 9999

Install
===========

git clone https://github.com/pyenv/pyenv.git ~/.pyenv

echo 'export PYENV_ROOT="$HOME/.pyenv"' >> ~/.bashrc

echo 'export PATH="$PYENV_ROOT/bin:$PATH"' >> ~/.bashrc

echo -e 'if command -v pyenv 1>/dev/null 2>&1; then\n  eval "$(pyenv init -)"\nfi' >> ~/.bashrc


virtualenv install:


git clone https://github.com/pyenv/pyenv-virtualenv.git $(pyenv root)/plugins/pyenv-virtualenv

echo 'eval "$(pyenv virtualenv-init -)"' >> ~/.bashrc

## install bz2 libs

sudo apt-get install zlib1g-dev libbz2-dev

## install python 3.6.4

pyenv install 3.6.4


## create pyenv virtualenv

pyenv virtualenv 3.6.4 whoisd-rwhois

## activate virtualenv

pyenv activate whoisd-rwhois

(whoisd-rwhois-consumer) user@host:  pip install -r requirements.txt



CHANGELOG.md
# [Changelog](https://github.com/teadur/whoisd-rwhois/releases)

## [0.0.2](https://github.com/teadur/whoisd-rwhois/compare/0.0.1...0.0.2)

* [1dc0cfc](https://github.com/teadur/whoisd-rwhois/commit/1dc0cfc) allow specifing api url
* [bcd294a](https://github.com/teadur/whoisd-rwhois/commit/bcd294a) Allow binding to IPv4/IPv6, actualy use threads
* [9e405df](https://github.com/teadur/whoisd-rwhois/commit/9e405df) configurable debug / listen ip
* [c7a0575](https://github.com/teadur/whoisd-rwhois/commit/c7a0575) fix indention
* [1e423de](https://github.com/teadur/whoisd-rwhois/commit/1e423de) refactor output generation to be more testable
* [274ea0d](https://github.com/teadur/whoisd-rwhois/commit/274ea0d) refactor to RwhoisRequest.changed()
* [1115b18](https://github.com/teadur/whoisd-rwhois/commit/1115b18) refactor admin/tech contacts output
* [f26b130](https://github.com/teadur/whoisd-rwhois/commit/f26b130) badges
* [ecea4bf](https://github.com/teadur/whoisd-rwhois/commit/ecea4bf) remove python 3.4
* [14f9543](https://github.com/teadur/whoisd-rwhois/commit/14f9543) less envs for travis
* [fcca83d](https://github.com/teadur/whoisd-rwhois/commit/fcca83d) travis virtualenv
* [47e125d](https://github.com/teadur/whoisd-rwhois/commit/47e125d) Python versions to travis
* [c7744c0](https://github.com/teadur/whoisd-rwhois/commit/c7744c0) pathlib to reqs
* [4c5a414](https://github.com/teadur/whoisd-rwhois/commit/4c5a414) drop python 2.7 from Travis
* [c468a6e](https://github.com/teadur/whoisd-rwhois/commit/c468a6e) cleanup
* [9ea0638](https://github.com/teadur/whoisd-rwhois/commit/9ea0638) include Python 3.6 in travis
* [35c97fb](https://github.com/teadur/whoisd-rwhois/commit/35c97fb) status test
* [ef95c39](https://github.com/teadur/whoisd-rwhois/commit/ef95c39) cleanup
* [678f302](https://github.com/teadur/whoisd-rwhois/commit/678f302) initial test for RwhoisRequest
* [c3089ef](https://github.com/teadur/whoisd-rwhois/commit/c3089ef) provide some runtime metrics
* [abd902a](https://github.com/teadur/whoisd-rwhois/commit/abd902a) move restful whois request to separate class
* [877d7f9](https://github.com/teadur/whoisd-rwhois/commit/877d7f9) whoisd v0.1 - import from dev
* [3292679](https://github.com/teadur/whoisd-rwhois/commit/3292679) Initial commit



tests/conftest.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
    Dummy conftest.py for whoisd_rwhois.

    If you don't know what this is for, just leave it empty.
    Read more about conftest.py under:
    https://pytest.org/latest/plugins.html
"""
from __future__ import print_function, absolute_import, division

# import pytest


tests/test_whoisd.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import pytest
from whoisd_rwhois.whoisd import RwhoisRequest

__author__ = "Georg Kahest"
__copyright__ = "Georg Kahest"
__license__ = "mit"

def test_RwhoisRequest():
    assert RwhoisRequest.make("nimi.ee", "test_thread") == "midagi"
    assert RwhoisRequest.status("internet.ee") == "ok (paid and in zone)"
    assert RwhoisRequest.print("name", "internet.ee") == b"name        internet.ee\n"
    assert RwhoisRequest.section("Tiitel") == b"\nTiitel\n"


tests/test_skeleton.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import pytest
from whoisd_rwhois.skeleton import fib

__author__ = "Georg Kahest"
__copyright__ = "Georg Kahest"
__license__ = "mit"


def test_fib():
    assert fib(1) == 1
    assert fib(2) == 1
    assert fib(7) == 13
    with pytest.raises(AssertionError):
        fib(-10)



src/whoisd_rwhois/skeleton.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
This is a skeleton file that can serve as a starting point for a Python
console script. To run this script uncomment the following line in the
entry_points section in setup.cfg:

    console_scripts =
     fibonacci = whoisd_rwhois.skeleton:run

Then run `python setup.py install` which will install the command `fibonacci`
inside your current environment.
Besides console scripts, the header (i.e. until _logger...) of this file can
also be used as template for Python modules.

Note: This skeleton file can be safely removed if not needed!
"""
from __future__ import division, print_function, absolute_import

import argparse
import sys
import logging
import socket
import threading
import socketserver
from whoisd_rwhois import __version__
from simple_rest_client.api import API
from simple_rest_client.resource import Resource

__author__ = "Georg Kahest"
__copyright__ = "Georg Kahest"
__license__ = "mit"

_logger = logging.getLogger(__name__)


class WhoisResource(Resource):
   actions = {
        'v1': {'method': 'GET', 'url': '/v1/{}.json'},
        'v2': {'method': 'GET', 'url': '/v2/{}.json'}

   }

class MetricData():
    finish = 0

class RwhoisRequest():
        
        kind = 'canine'         # class variable shared by all instances

        def __init__(self, name):
            self.name = name    # instance variable unique to each instance
                        
            _logger.debug('RwhoisRequest:')
            _logger.debug(self.name)

        def make(domain_name,cur_thread):
            _logger.debug('RwhoisRequest.make:')
            _logger.debug(domain_name)

            # restful-whois
            default_params = {}
            rwhois_api = API(
                api_root_url='https://rwhois.internet.ee/', params=default_params,
                json_encode_body=True
            )
            rwhois_api.add_resource(resource_name='domain', resource_class=WhoisResource)
            rwhois_response = rwhois_api.domain.v1(domain_name, body=None, params={}, headers={})
            # rwhois_response = rwhois_api.domain.v2(domain_name, body=None, params={'access_token': 'valid-token'}, headers={})

            sisu = rwhois_response.body
            response_time = rwhois_response.headers['X-Runtime']
            response_date = rwhois_response.headers['Date']

            # TODO: implement discolese  in rwhois json?
            sisu['disclose'] = "Not Disclosed - Visit www.internet.ee for webbased WHOIS"

            response = bytes("{}: {} {} time: {} @ {}\n".format(cur_thread.name, domain_name, MetricData.finish, response_time, response_date), 'utf-8')
            response += bytes("Estonia .ee Top Level Domain WHOIS server \n \nDomain: \n", 'utf-8')
            response += bytes("{:<12}{}\n".format('name:', sisu['name']), 'utf-8')

            # handle multiple status entries
            for x in sisu['status']:
                response += bytes("{:<12}{}\n".format('status:', x), 'utf-8')

            response += bytes(
                "{:<12}{}\n".format('registered:', sisu['registered'].replace("T", " ").replace("+", " +")), 'utf-8')
            response += bytes("{:<12}{}\n".format('changed:', sisu['changed'].replace("T", " ").replace("+", " +")),
                              'utf-8')
            response += bytes("{:<12}{}\n".format('expire:', sisu['expire']), 'utf-8')
            response += bytes("{:<12}{}\n".format('outzone:', str(sisu['outzone'] or '')), 'utf-8')
            response += bytes("{:<12}{}\n\n".format('delete:', str(sisu['delete'] or '')), 'utf-8')

            # Registrant
            response += bytes("{}\n{:<12}{}\n".format('Registrant:', 'name:', sisu['registrant']), 'utf-8')
            if sisu['registrant_kind'] == "org":
                response += bytes("{:<12}{}\n".format('org id:', sisu['registrant_reg_no']), 'utf-8')
                response += bytes("{:<12}{}\n".format('country:', sisu['registrant_ident_country_code']), 'utf-8')
            response += bytes("{:<12}{}\n".format('email:', sisu['disclose']), 'utf-8')
            response += bytes(
                "{:<12}{}\n".format('changed:', sisu['registrant_changed'].replace("T", " ").replace("+", " +")),
                'utf-8')

            # Administrative contacts
            response += bytes("\n{}\n".format('Administrative contact:'), 'utf-8')
            for contact in sisu['admin_contacts']:
                response += bytes("{:<12}{}\n".format('name:', contact['name']), 'utf-8')
                response += bytes("{:<12}{}\n".format('email:', sisu['disclose']), 'utf-8')
                response += bytes("{:<12}{}\n".format('changed:', contact['changed'].replace("T", " ").replace("+", " +")),
                                  'utf-8')

            # Tech contacs
            response += bytes("\n{}\n".format('Technical contact:'), 'utf-8')
            for contact in sisu['tech_contacts']:
                response += bytes("{:<12}{}\n".format('name:', contact['name']), 'utf-8')
                response += bytes("{:<12}{}\n".format('email:', sisu['disclose']), 'utf-8')
                response += bytes("{:<12}{}\n".format('changed:', contact['changed'].replace("T", " ").replace("+", " +")),
                                  'utf-8')

            # Registrar
            response += bytes("\n{}\n".format('Registrar:'), 'utf-8')
            response += bytes("{:<12}{}\n".format('name:', sisu['registrar']), 'utf-8')
            response += bytes("{:<12}{}\n".format('url:', sisu['registrar_website']), 'utf-8')
            response += bytes("{:<12}{}\n".format('phone:', sisu['registrar_phone']), 'utf-8')
            response += bytes(
                "{:<12}{}\n".format('changed:', sisu['registrar_changed'].replace("T", " ").replace("+", " +")),
                'utf-8')

            # Name servers
            response += bytes("\n{}\n".format('Name servers:'), 'utf-8')
            for nameserver in sisu['nameservers']:
                response += bytes("{:<12}{}\n".format('nserver:', nameserver), 'utf-8')
            response += bytes(
                "{:<12}{}\n".format('changed:', sisu['nameservers_changed'].replace("T", " ").replace("+", " +")),
                'utf-8')

            # DNSSec Keys
            if sisu['dnssec_keys'] != []:
                response += bytes("\n{}\n".format('DNNSEC:'), 'utf-8')
                for key in sisu['dnssec_keys']:
                    response += bytes("{:<12}{}\n".format('dnskey:', key), 'utf-8')
                response += bytes(
                    "{:<12}{}\n".format('changed:', sisu['dnssec_changed'].replace("T", " ").replace("+", " +")),
                    'utf-8')

            # Footer
            response += bytes("\nEstonia .ee Top Level Domain WHOIS server\nMore information at https://internet.ee",
                              'utf-8')

            return response

class ThreadedTCPRequestHandler(socketserver.BaseRequestHandler):


    def handle(self):

        # take data in as byte input
        data = self.request.recv(1024).strip()

        # convert input to utf-8 string
        data_str = str(data, 'utf-8')
        cur_thread = threading.current_thread()
        _logger.debug("thread: ")
        _logger.debug(cur_thread)


        response = RwhoisRequest.make(data_str,cur_thread)

        # try:
        #    rwhois_response = rwhois_api.domain.whois(data_str, body=None, params={}, headers={})
        # except Exception as e:
        #    z = e  # representation: "<exceptions.ZeroDivisionError instance at 0x817426c>"
        #    _logger.debug(z)  # output: "integer division or modulo by zero"

        # Send response
        self.request.sendall(response)

class ThreadedTCPServer(socketserver.ThreadingMixIn, socketserver.TCPServer):
    def __init__(self, server_address,
                 handler_class=ThreadedTCPRequestHandler,
                 ):
        self.logger = logging.getLogger('ThreadedTCPServer')
        self.logger.debug('__init__')
        socketserver.TCPServer.__init__(self, server_address,
                                        handler_class)
        return

    def server_activate(self):
        self.logger.debug('server_activate')
        socketserver.TCPServer.server_activate(self)
        return

    def serve_forever(self, poll_interval=0.5):
        self.logger.debug('waiting for request')
        self.logger.info(
            'Handling requests, press <Ctrl-C> to quit'
        )
        socketserver.TCPServer.serve_forever(self, poll_interval)
        return

    def handle_request(self):
        self.logger.debug('handle_request')
        return socketserver.TCPServer.handle_request(self)

    def verify_request(self, request, client_address):
        self.logger.debug('verify_request(%s, %s)',
                          request, client_address)
        return socketserver.TCPServer.verify_request(
            self, request, client_address,
        )

    def process_request(self, request, client_address):
        self.logger.debug('process_request(%s, %s)',
                          request, client_address)

        return socketserver.TCPServer.process_request(
            self, request, client_address,
        )

    def server_close(self):
        self.logger.debug('server_close')
        return socketserver.TCPServer.server_close(self)

    def finish_request(self, request, client_address):
        self.logger.debug('finish_request(%s, %s)',
                          request, client_address)
        MetricData.finish+=1
        self.logger.debug(MetricData.finish)
        return socketserver.TCPServer.finish_request(
            self, request, client_address,
        )

    def close_request(self, request_address):
        self.logger.debug('close_request(%s)', request_address)
        return socketserver.TCPServer.close_request(
            self, request_address,
        )

    def shutdown(self):
        self.logger.debug('shutdown()')
        return socketserver.TCPServer.shutdown(self)


def client(ip, port, message):
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.connect((ip, port))
    try:
        sock.sendall(bytes(message, 'ascii'))
        response = str(sock.recv(1024), 'ascii')
        print("Received: {}".format(response))
    finally:
        sock.close()


def fib(n):
    """Fibonacci example function

    Args:
      n (int): integer

    Returns:
      int: n-th Fibonacci number
    """
    assert n > 0
    a, b = 1, 1
    for i in range(n-1):
        a, b = b, a+b
    return a


def parse_args(args):
    """Parse command line parameters

    Args:
      args ([str]): command line parameters as list of strings

    Returns:
      :obj:`argparse.Namespace`: command line parameters namespace
    """
    parser = argparse.ArgumentParser(
        description="Just a Fibonnaci demonstration")
    parser.add_argument(
        '--version',
        action='version',
        version='whoisd-rwhois {ver}'.format(ver=__version__))
    parser.add_argument(
        dest="n",
        help="n-th Fibonacci number",
        type=int,
        metavar="INT")
    parser.add_argument(
        '-v',
        '--verbose',
        dest="loglevel",
        help="set loglevel to INFO",
        action='store_const',
        const=logging.INFO)
    parser.add_argument(
        '-vv',
        '--very-verbose',
        dest="loglevel",
        help="set loglevel to DEBUG",
        action='store_const',
        const=logging.DEBUG)
    return parser.parse_args(args)


def setup_logging(loglevel):
    """Setup basic logging

    Args:
      loglevel (int): minimum loglevel for emitting messages
    """
    logformat = "[%(asctime)s] %(levelname)s:%(name)s:%(message)s"
    logging.basicConfig(level=loglevel, stream=sys.stdout,
                        format=logformat, datefmt="%Y-%m-%d %H:%M:%S")


def main(args):
    """Main entry point allowing external calls

    Args:
      args ([str]): command line parameter list
    """
    args = parse_args(args)
    setup_logging(args.loglevel)
    _logger.debug("Starting crazy calculations...")
    print("The {}-th Fibonacci number is {}".format(args.n, fib(args.n)))
    HOST, PORT = "localhost", 9999

    server = ThreadedTCPServer((HOST, PORT), ThreadedTCPRequestHandler)
    ip, port = server.server_address

    # Start a thread with the server -- that thread will then start one
    # more thread for each request
    server_thread = threading.Thread(target=server.serve_forever)
    # Exit the server thread when the main thread terminates
    server_thread.daemon = True
    server_thread.start()
    print("Server loop running in thread:", server_thread.name)

    # client(ip, port, "Hello World 1")
    # client(ip, port, "Hello World 2")
    # client(ip, port, "Hello World 3")

    # start server
    server.serve_forever()
    # rerver.server_activate()


    _logger.info("Script ends here")


def run():
    """Entry point for console_scripts
    """
    main(sys.argv[1:])


if __name__ == "__main__":
    run()


src/whoisd_rwhois/whoisd.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
Unix whois daemon with restful-whois backend.

"""
# from __future__ import division, print_function, absolute_import

import argparse
import sys
import logging
import socket
import threading
import socketserver
from whoisd_rwhois import __version__
from simple_rest_client.api import API
from simple_rest_client.resource import Resource

__author__ = "Georg Kahest"
__copyright__ = "Georg Kahest"
__license__ = "mit"

_logger = logging.getLogger(__name__)


class WhoisResource(Resource):
   actions = {
        'v1': {'method': 'GET', 'url': '/v1/{}.json'},
        'v2': {'method': 'GET', 'url': '/v2/{}.json'}

   }

class MetricData():
    finish = 0
    test_request = 0

class RwhoisRequest():

    # TODO: implement discolese  in rwhois json?
        disclosed = "Not Disclosed - Visit www.internet.ee for webbased WHOIS"

        debug_ip = 0
        debug_src = ""
        ipversion = socket.AF_INET
        api_url = "https://rwhois.internet.ee/"

        def __init__(self, name):
            self.name = name
            _logger.debug("RwhoisRequest: init {}".format(self.name))


        def get(domain_name):

            default_params = {}
            rwhois_api = API(
                api_root_url=RwhoisRequest.api_url, params=default_params,
                json_encode_body=True
            )

            rwhois_api.add_resource(resource_name='domain', resource_class=WhoisResource)
            rwhois_response = rwhois_api.domain.v1(domain_name, body=None, params={}, headers={})

            return rwhois_response

        def status(domain_name):
            return RwhoisRequest.get(domain_name).body['status'][0]

        def contacts(contacts):

            for contact in contacts:
                response = bytes("{:<12}{}\n".format('name:', contact['name']), 'utf-8')
                response += RwhoisRequest.print('email', RwhoisRequest.disclosed)
                response += RwhoisRequest.changed(contact['changed'])

            return response

        def changed(changed):

            response = bytes("{:<12}{}\n".format('changed:', changed.replace("T", " ").replace("+", " +")), 'utf-8')
            return response

        def print(key, value):

            response = bytes("{:<12}{}\n".format(key, value), 'utf-8')
            return response

        def section(title):

            response = bytes("\n{}\n".format(title), 'utf-8')
            return response

        def make(domain_name,cur_thread):
            thread_name = str(cur_thread or cur_thread.name)

            _logger.debug('RwhoisRequest.make: {}'.format(domain_name))

            rwhois_response = RwhoisRequest.get(domain_name)
            # rwhois_response = rwhois_api.domain.v2(domain_name, body=None, params={'access_token': 'valid-token'}, headers={})

            sisu = rwhois_response.body
            response_time = rwhois_response.headers['X-Runtime']
            response_date = rwhois_response.headers['Date']

            response = b""
            _logger.debug("Debug_ip:{}".format(RwhoisRequest.debug_ip))
            # Metrics line
            if RwhoisRequest.debug_ip == 1:
                response += bytes("{}: {} {} time: {} @ {}\n".format(thread_name, domain_name, MetricData.finish, response_time, response_date), 'utf-8')

            response += bytes("Estonia .ee Top Level Domain WHOIS server \n \nDomain: \n", 'utf-8')
            response += RwhoisRequest.print('name:',sisu['name'])

            # handle multiple status entries
            for status in sisu['status']:
                 response += RwhoisRequest.print('status',status)

            response += bytes("{:<12}{}\n".format('registered:', sisu['registered'].replace("T", " ").replace("+", " +")), 'utf-8')
            response += RwhoisRequest.changed(sisu['changed'])
            response += RwhoisRequest.print('expire:',sisu['expire'])
            response += RwhoisRequest.print('outzone:', str(sisu['outzone'] or ''))
            response += RwhoisRequest.print('delete:', str(sisu['delete'] or ''))


            # Registrant
            response += RwhoisRequest.section('Registrant:')
            response += RwhoisRequest.print('name:', sisu['registrant'])

            if sisu['registrant_kind'] == "org":
                response += RwhoisRequest.print('org id:',sisu['registrant_reg_no'])
                response += RwhoisRequest.print('country:',sisu['registrant_ident_country_code'])
            response += RwhoisRequest.print('email', RwhoisRequest.disclosed)
            response += RwhoisRequest.changed(sisu['registrant_changed'])


            # Administrative contacts
            response += RwhoisRequest.section('Administrative contact:')
            response += RwhoisRequest.contacts(sisu['admin_contacts'])

            # Tech contacs
            response += RwhoisRequest.section('Technical contact:')
            response += RwhoisRequest.contacts(sisu['tech_contacts'])

            # Registrar
            response += RwhoisRequest.section('Registrar:')
            response += RwhoisRequest.print('name:',sisu['registrar'])
            response += RwhoisRequest.print('url:',sisu['registrar_website'])
            response += RwhoisRequest.print('phone:',sisu['registrar_phone'])
            response += RwhoisRequest.changed(sisu['registrar_changed'])

            # Name servers
            response += bytes("\n{}\n".format('Name servers:'), 'utf-8')
            for nameserver in sisu['nameservers']:
                response += RwhoisRequest.print('nserver:', nameserver)

            response += RwhoisRequest.changed(sisu['nameservers_changed'])

            # DNSSec Keys
            if sisu['dnssec_keys'] != []:
                response += bytes("\n{}\n".format('DNNSEC:'), 'utf-8')
                for key in sisu['dnssec_keys']:
                    response += RwhoisRequest.print('dnskey:', key)
                response += RwhoisRequest.changed(sisu['dnssec_changed'])

            # Footer
            response += bytes("\nEstonia .ee Top Level Domain WHOIS server\nMore information at https://internet.ee",
                              'utf-8')

            # Ugly hack to run tests
            if thread_name == "test_thread":
               _logger.debug(sys.getsizeof(response))
               if sys.getsizeof(response) == 1118:
                    return "midagi"

            return response

class ThreadedTCPRequestHandler(socketserver.BaseRequestHandler):


    def handle(self):

        # take data in as byte input
        data = self.request.recv(1024).strip()

        # convert input to utf-8 string
        data_str = str(data, 'utf-8')
        cur_thread = threading.current_thread()
        _logger.debug("thread:  {}".format(cur_thread))
        _logger.debug("client:  {}".format(self.client_address))


        response = RwhoisRequest.make(data_str,cur_thread)
        # Send response
        self.request.sendall(response)

class ThreadedTCPServer(socketserver.ThreadingMixIn, socketserver.TCPServer):

    def __init__(self, server_address,
                 handler_class=ThreadedTCPRequestHandler,
                 ):
        self.allow_reuse_address = True
        self.address_family = RwhoisRequest.ipversion
        self.request_queue_size=50
        self.logger = logging.getLogger('ThreadedTCPServer')
        self.logger.debug('__init__')
        socketserver.TCPServer.__init__(self, server_address,
                                        handler_class)
        return

    def server_activate(self):
        self.logger.debug('server_activate')
        socketserver.TCPServer.server_activate(self)
        return

    def serve_forever(self, poll_interval=0.5):
        self.logger.debug('waiting for request')
        self.logger.info(
            'Handling requests, press <Ctrl-C> to quit'
        )
        socketserver.TCPServer.serve_forever(self, poll_interval)
        return

    def handle_request(self):
        self.logger.debug('handle_request')
        return socketserver.TCPServer.handle_request(self)

    def verify_request(self, request, client_address):
        self.logger.debug('verify_request(%s, %s)',
                          request, client_address)
        # TODO: configurable debug ip dict
        if client_address[0] == RwhoisRequest.debug_src:
            _logger.debug("request from debug ip")
            RwhoisRequest.debug_ip=1
        else:
            RwhoisRequest.debug_ip=0


        return socketserver.TCPServer.verify_request(
            self, request, client_address,
        )

    def process_request(self, request, client_address):
        self.logger.debug('process_request(%s, %s)',
                          request, client_address)

        return socketserver.TCPServer.process_request(
            self, request, client_address,
        )

    def server_close(self):
        self.logger.debug('server_close')
        return socketserver.TCPServer.server_close(self)

    def finish_request(self, request, client_address):
        self.logger.debug('finish_request(%s, %s)',
                          request, client_address)
        MetricData.finish+=1
        self.logger.debug(MetricData.finish)
        return socketserver.TCPServer.finish_request(
            self, request, client_address,
        )

    def close_request(self, request_address):
        self.logger.debug('close_request(%s)', request_address)
        return socketserver.TCPServer.close_request(
            self, request_address,
        )

    def shutdown(self):
        self.logger.debug('shutdown()')
        return socketserver.TCPServer.shutdown(self)

def parse_args(args):
    """Parse command line parameters

    Args:
      args ([str]): command line parameters as list of strings

    Returns:
      :obj:`argparse.Namespace`: command line parameters namespace
    """
    parser = argparse.ArgumentParser(
        description="Unix Whois Daemon for Restful Whois backend")
    parser.add_argument(
        '--version',
        action='version',
        version='whoisd-rwhois {ver}'.format(ver=__version__))
    parser.add_argument(
        dest="p",
        help="Port to listen on",
        type=int,
        metavar="INT")

    parser.add_argument(
        '-4',
        dest="ip",
        help="Bind to IPv4",
        action='store_const',
        const=socket.AF_INET)

    parser.add_argument(
        '-6',
        dest="ip",
        help="Bind to IPv6",
        action='store_const',
        const=socket.AF_INET6)

    parser.add_argument(
        '-l',
        '--listen',
        '--host',
        dest="h",
        help="host to listen on",
        metavar="127.0.0.1",
        default="127.0.0.1")
    parser.add_argument(
        '-u',
        '--url',
        dest="url",
        help="Backend API url",
        metavar="https://rwhois.internet.ee/",
        default="https://rwhois.internet.ee/")
    parser.add_argument(
        '-d',
        '--debug',
        dest="d",
        help="debug ip",
        metavar="127.0.0.1",
        default="127.0.0.1")
    parser.add_argument(
        '-v',
        '--verbose',
        dest="loglevel",
        help="set loglevel to INFO",
        action='store_const',
        const=logging.INFO)
    parser.add_argument(
        '-vv',
        '--very-verbose',
        dest="loglevel",
        help="set loglevel to DEBUG",
        action='store_const',
        const=logging.DEBUG)
    return parser.parse_args(args)


def setup_logging(loglevel):
    """Setup basic logging

    Args:
      loglevel (int): minimum loglevel for emitting messages
    """
    logformat = "[%(asctime)s] %(levelname)s:%(name)s:%(message)s"
    logging.basicConfig(level=loglevel, stream=sys.stdout,
                        format=logformat, datefmt="%Y-%m-%d %H:%M:%S")


def main(args):
    """Main entry point allowing external calls

    Args:
      args ([str]): command line parameter list
    """
    args = parse_args(args)
    setup_logging(args.loglevel)
    RwhoisRequest.debug_src=args.d
    RwhoisRequest.ipversion=args.ip
    RwhoisRequest.api_url=args.url
    HOST, PORT = args.h, args.p
    server = ThreadedTCPServer((HOST, PORT), ThreadedTCPRequestHandler)
    # ip, port = server.server_address
    # server.address_family=(socket.AF_INET6)

    # Start a thread with the server -- that thread will then start one
    # more thread for each request
    server_thread = threading.Thread(target=server.serve_forever())
    # Exit the server thread when the main thread terminates
    server_thread.daemon = True
    server_thread.start()
    print("Server loop running in thread:", server_thread.name)

    _logger.info("Script ends here")


def run():
    """Entry point for console_scripts
    """
    main(sys.argv[1:])


if __name__ == "__main__":
    run()
